x-pack/plugins/apm/server/lib/helpers/get_document_sources.ts

```diff
@@ -64,170 +78,169 @@ export async function getDocumentSources({
-    current,
-  ]);
+  const allSearches = sourceRequests
+    .flatMap(({ before, current, durationSummaryCheck }) => [
+      before,
+      current,
+      durationSummaryCheck,
+    ])
+    .filter(
+      (request): request is ReturnType<typeof getRequest> =>
+        request !== undefined
+    );
 
   const allResponses = (
     await apmEventClient.msearch('get_document_availability', ...allSearches)
   ).responses;
 
-  const checkedSources = allSourcesToCheck.map((source, index) => {
-    const { documentType, rollupInterval } = source;
-    const responseBefore = allResponses[index * 2];
-    const responseAfter = allResponses[index * 2 + 1];
-
-    const hasDocBefore = responseBefore.hits.total.value > 0;
-    const hasDocAfter = responseAfter.hits.total.value > 0;
+  return sourceRequests.map(({ documentType, rollupInterval, ...queries }) => {
+    const numberOfQueries = Object.values(queries).filter(Boolean).length;
+    // allResponses is sorted by the order of the requests in sourceRequests
+    const docTypeResponses = allResponses.splice(0, numberOfQueries);
 
     return {
       documentType,
       rollupInterval,
-      hasDocBefore,
-      hasDocAfter,
-      checkSummaryFieldExists: source.meta.checkSummaryFieldExists,
+      hasDocBefore: docTypeResponses[QUERY_INDEX.BEFORE].hits.total.value > 0,
+      hasDocAfter: docTypeResponses[QUERY_INDEX.CURRENT].hits.total.value > 0,
+      hasDurationSummary: docTypeResponses[QUERY_INDEX.DURATION_SUMMARY]
+        ? docTypeResponses[QUERY_INDEX.DURATION_SUMMARY].hits.total.value === 0
+        : true,
     };
   });
+};
 
-  const hasAnySourceDocBefore = checkedSources.some(
-    (source) => source.hasDocBefore
-  );
+const getDocumentTypeRequestsFn =
+  ({
+    enableContinuousRollups,
+    start,
+    end,
+    kuery,
+  }: {
+    enableContinuousRollups: boolean;
+    start: number;
+    end: number;
+    kuery: string;
+  }) =>
+  (documentType: ApmDocumentType) => {
+    const currentRange = rangeQuery(start, end);
+    const diff = end - start;
+    const kql = kqlQuery(kuery);
+    const beforeRange = rangeQuery(start - diff, end - diff);
+
+    const rollupIntervals = enableContinuousRollups
+      ? getConfigForDocumentType(documentType).rollupIntervals
+      : [RollupInterval.OneMinute];
+
+    return rollupIntervals.map((rollupInterval) => ({
+      documentType,
+      rollupInterval,
+      before: getRequest({
+        documentType,
+        rollupInterval,
+        filters: [...kql, ...beforeRange],
+      }),
+      current: getRequest({
+        documentType,
+        rollupInterval,
+        filters: [...kql, ...currentRange],
+      }),
+      ...(documentType !== ApmDocumentType.ServiceTransactionMetric
+        ? {
+            durationSummaryCheck: getRequest({
+              documentType,
+              rollupInterval,
+              filters: [...kql, ...currentRange, getDurationLegacyFilter()],
+            }),
+          }
+        : {}),
+    }));
```

Comment:
Should we not check both before/after here?

Reply:
I don't think we need to - but I might be wrong. The `hasDoc` case seems to be different, at least based on the comment provided. For the `durationSummary` I think we're only interested if in the current timestamp all documents have the new field.

Reply:
I think you're right! Because the check is inversed (which is not possible with document sources), checking the current time range is enough.